[2단계 - 페이먼츠] 기린(정유정) 미션 제출합니다.

jeongyou opened the pull request:
안녕하세요 유조! 기린입니다.  
해외 일정은 무사히 잘 다녀오셨나요?
Step2 코드 리뷰도 잘 부탁드립니다! 🙇‍♀️

## 🎯 페이먼츠

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계
- 재사용 가능한 Input Component를 개발한다.
- Storybook을 사용하여 컴포넌트의 다양한 상태를 시각적으로 테스트한다.
- 카드 정보를 효과적으로 렌더링 하기 위한 상태 관리를 경험한다.

### 2단계
- 다양한 Form 구성 요소들간의 상태를 효율적으로 관리한다.
- hooks API를 이용하여 상태 관리 로직을 구현하한다.
- custom hooks를 생성하여 Form 관리 로직을 컴포넌트에서 분리하고 재사용한다.
- Controlled & Uncontrolled Components에 입각하여 Form을 핸들링한다.

---

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 기본적인 프로그래밍 요구 사항(코드 컨벤션, 에러 핸들링 등)을 준수했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 기본 배포 링크 기입: [https://jeongyou.github.io/react-payments/](https://jeongyou.github.io/react-payments/)
  - 스토리북 링크: [https://67ff7d17e8344e018910c187-sepasgnwoz.chromatic.com/?path=/docs/components-cvcnumbers--docs](https://67ff7d17e8344e018910c187-sepasgnwoz.chromatic.com/?path=/docs/components-cvcnumbers--docs)

### 리뷰 요청 & 논의하고 싶은 내용

#### 1) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점
**1. 컴포넌트의 크기와 분리 기준**
이번 미션에서는 컴포넌트를 어떻게 분리할지 고민했습니다.
상향식 설계를 위해 HTML 태그 수준의 기본 컴포넌트를 먼저 만들고, 이를 각 컴포넌트에서 커스터마이징하여 사용하도록 설계했습니다. 그래서 작은 컴포넌트들이 결합되어 하나의 화면을 구성하도록 했습니다.

**2. Domain과 UI 분리**
Domain과 UI를 분리하기 위해 Custom Hook을 처음으로 활용해보았습니다.
처음에는 Custom Hook이 "중복 로직을 추상화하는 도구" 정도로만 이해했지만, 이번에 더 찾아보면서 "UI와 Domain을 분리하는 역할"을 한다는 것을 알게 되었습니다. 이에 따라 Form 상태 관리 로직을 Custom Hook으로 분리하여, 컴포넌트는 UI 렌더링에만 집중할 수 있도록 설계했습니다.

**3. 카드 정보 입력 방향 (거꾸로 입력)**
입력 필드를 추가하는 방식에 대해 고민했습니다. 하단에 입력창을 추가하면 사용자가 매번 스크롤을 내려야 해서 불편할 것 같아, 입력이 완료되면 위쪽에 새로운 입력 필드가 생성되도록 UX를 개선해보았습니다.

**4. 입력 필드 자동 이동**
사용자가 입력을 완료하면 다음 입력 필드로 자동 포커스가 이동하도록 구현했습니다.
이를 통해 사용자가 키보드나 마우스를 별도로 조작하지 않고도 편리하게 입력을 이어갈 수 있도록 UX를 향상시켰습니다.

**5. react-router를 이용한 페이지 관리**
  - 카드 정보 입력 페이지와 등록 완료 페이지를 나누어 react-router로 관리했습니다.
  - `useNavigate`를 이용해 카드 등록 완료 시 카드 번호 앞 4자리와 카드 브랜드 정보를 state로 전달했습니다. 등록 완료 페이지에서는 `useLocation`을 통해 이 state를 받아 사용했습니다.
  - `window.history.replaceState(null, '', window.location.href);`를 통해 뒤로 가기 후 앞으로 가기 시에도 state가 남지 않도록 처리했습니다.
  - 등록 완료 페이지에 직접 접근했을 때(state가 없는 경우) 카드 정보 입력 페이지로 리다이렉트하도록 설정해, 예외 상황도 처리했습니다.

**6. Storybook 활용**
Storybook의 Docs 기능을 활용해, 각 컴포넌트와 페이지의 기능 및 역할을 명확히 설명해서 나중에 다른 사람들과 협업할 때, 컴포넌트 사용법이나 역할을 쉽게 이해할 수 있도록 문서를 보완했습니다.

#### 2) 이번 리뷰를 통해 논의하고 싶은 부분
**1. 기본 컴포넌트 생성의 필요성**
HTML 태그 수준의 기본 컴포넌트를 만드는 것이 과연 의미가 있는지 고민이 들었습니다.
예를 들어서 저는 SingleButton이라는 기본 컴포넌트를 만들고, 이를 바탕으로 두 개의 서로 다른 기능을 가진 '확인 버튼'을 만들었는데, 
- "기본 컴포넌트를 만들어서 재사용하는 것"과 
- "각 컴포넌트에서 그냥 HTML 태그를 바로 사용하는 것"
둘의 실질적인 차이를 잘 체감하지 못했습니다.
그서 현재 제가 작성한 방식처럼 HTML 태그 수준의 기본 컴포넌트를 만드는 것에 대한 유조의 의견이 궁금합니다!

**2. 커스텀 훅 사용**
이번 미션을 하면서 Custom Hook을 활용해 UI와 Domain 코드를 분리하는 장점을 확실히 체감했습니다.
하지만 컴포넌트별로 도메인 로직을 분리하다 보니, 기능이 조금만 달라도 커스텀 훅을 새로 만들어야 하는 상황이 많아졌습니다.
중복되는 로직이 있다면 재사용할 수 있지만, 실제로는 약간씩 다른 도메인 로직이 많아 커스텀 훅이 계속 늘어나는 느낌이 들었습니다.
그래서 컴포넌트가 추가될 때마다 커스텀 훅을 계속 추가하는 것이 맞는지, 혹은 더 좋은 방법이 있는지 유조의 생각이 궁금합니다.

**3. 입력 필드 자동 이동**
현재 각 입력 컴포넌트 안에서는 입력이 완료되면 내부 필드 간 포커스 이동이 되지만, 한 컴포넌트의 입력이 끝난 후 다음 컴포넌트로 포커스가 넘어가는 기능은 구현하지 않았습니다.
처음에는 입력 완료 후 다음 컴포넌트로 자동 이동하는 것도 고려했지만, 연속 입력할 때는 편리한 반면, 중간 필드에서 오류가 발생했을 경우 사용자가 오류를 고친 후 포커스가 이동해버리면 다시 원하는 필드를 클릭해야 하는 번거로움이 생길 수 있다고 생각했습니다.
이러한 상황을 고려하여, 이번 미션에서는 사용자가 필요한 경우 직접 이동하는 흐름을 선택했습니다.
유조는 입력 필드 자동이동을 할 때 어떤 방식을 주로 사용하시나요? 이러한 UX 선택에 대해 유조의 의견이 궁금합니다!

**4. react-router를 이용한 페이지 관리**
등록 완료 후 뒤로 가기를 눌렀다가 앞으로 가기를 누르면, 다시 등록 완료 페이지에 접근할 수 있는 문제가 있었습니다. 그래거 이를 해결하기 위해, state가 없는 경우(카드 정보를 모두 입력하고 완료 버튼을 눌렀을 때만 존재하는 state) 카드 입력 페이지로 리다이렉트하는 처리를 추가했습니다. 예외 상황을 막기 위한 방법이였는데 한편으로는 사용자의 앞/뒤로 가기 동작을 강제로 막는 것이 과연 좋은 방법일까? 라는 고민도 들었습니다. 개인적으로 리다이렉트를 막으면서 상황을 안내하는 alert를 띄우는 방법도 고려해볼 수 있지 않을까 생각이 듭니다.
유조는 이런 페이지 접근 제어를 할 때 보통 어떤 방법을 선택하시나요? 또, 현재와 같은 리다이렉트 방식에 대한 의견도 듣고 싶습니다!

이상입니다!  
이번 미션을 진행하면서 고민하고 배운 부분들을 최대한 담아보았습니다.  
피드백 주시면 더 고민하고 개선해보겠습니다. 🙇‍♀️  
오늘도 좋은 하루 보내세요!

---

## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도만이 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. Form 상태 관리 & Custom Hook 분리

- 반복되는 로직을 custom hook으로 분리했는가?
- hook 내부와 UI 컴포넌트의 **역할이 명확하게 분리**되어 있는가?
- 상태 흐름이 직관적이며, Form 전체를 일관되게 관리할 수 있는 구조인가?

### 2. 입력 UI 흐름과 UX

- 카드 번호 입력 시 필드 간 **자동 포커싱 이동**이 자연스럽게 동작하는가?
- 숫자만 입력 가능한 필드에서 제한, 에러 메시지, 유효성 피드백 등 **사용자 경험이 충분히 고려**되었는가?

### 3. 컴포넌트 구조 및 재사용성

- 컴포넌트가 명확한 **역할과 책임**을 가지며 과도하게 분리되거나 중첩되어 있지 않은가?
- Input, Button 등 재사용 가능한 컴포넌트가 잘 정의되어 있는가?

### 4. 상태 기반 유효성 검사 및 확인 버튼 활성화

- 모든 필드가 유효할 때만 확인 버튼이 **정확히 활성화/비활성화**되는가?
- 유효성 검사의 기준이 명확하고, **상태 변경에 따른 UI 반응**이 잘 연결되어 있는가?

jeongyou committed:
refactor: 중복된 에러 상태 로직을 useErrorMessage 훅으로 추출

jeongyou committed:
docs: step1의 기능 구현 목록

jeongyou committed:
docs: step2 기능 구현 목록 작성

jeongyou committed:
docs: 기능 구현 목록 수정

jeongyou committed:
refactor: InputLabel 컴포넌트를 FormLabel로 이름 변경

- Input 외에도 Select 등 다양한 폼 요소에서 사용할 수 있도록 범용적인 이름인 FormLabel로 변경

jeongyou committed:
feat: 카드사 선택 컴포넌트 구현

jeongyou committed:
feat: 카드사에 따라 미리보기 카드의 색상을 변경하는 기능 구현

jeongyou committed:
docs: 카드사 선택 기능 구현 완료

jeongyou committed:
feat: 비밀번호 앞 2자리 입력 컴포넌트 구현

jeongyou committed:
feat: 동적으로 input의 포커스를 이동, 모든 입력 완료 시 함수를 실행하는 커스텀 훅 구현

jeongyou committed:
feat: 하나의 입력 그룹의 모든 입력이 완료 되면, 동적으로 다음 입력 그룹이 나타나는 기능 구현

jeongyou committed:
feat: 선택된 카드 브랜드를 저장하는 기능 구현

jeongyou committed:
chore: 불필요한 import 삭제

jeongyou committed:
feat: SingleInput 컴포넌트에 type props로 받도록 설정

jeongyou committed:
feat: 'text' | 'password' 를 input의 type속성 props로 넘겨주도록 구현

jeongyou committed:
feat: InputTexts의 props로 타입값을 넘겨주도록 구현

jeongyou committed:
feat: CVC 번호가 모두 입력되면 비밀번호 입력 컴포넌트가 화면에 나타나는 기능 구현

jeongyou committed:
docs: 구현 완료 기능 체크

jeongyou committed:
chore: 코드에 프리티어 적용됨

jeongyou committed:
style: 카드사 선택 컴포넌트의 밑 부분에 간격 추가

jeongyou committed:
feat: 기본 버튼 컴포넌트 구현

jeongyou committed:
feat: 폼 제출 버튼 컴포넌트 UI 구현 완료

jeongyou committed:
feat: 다시 카드 등록 페이지로 돌아갈 수 있는 확인 버튼 컴포넌트의 UI 구현

jeongyou committed:
chore: prettier 적용

jeongyou committed:
feat: 재사용 가능한 기본 폼 컴포넌트 생성

jeongyou committed:
feat: 카드 정보를 등록하는 폼 컴포넌트 구현

jeongyou committed:
chore: react-router-dom 라이브러리 설치

jeongyou committed:
refactor: onSubmit props를 선택적으로 처리할 수 있도록 변경

jeongyou committed:
refactor: 카드 등록 관련 화면을 별도 페이지로 분리

jeongyou committed:
feat: 기본 Image 컴포넌트 추가

jeongyou committed:
feat: 카드 등록 완료 페이지 UI 구현 완료

jeongyou committed:
chore: 카드 등록 완료 아이콘 이미지 추가

jeongyou committed:
chore: 불필요한 쉼표 삭제

jeongyou committed:
feat: 카드 정보 입력 완료 버튼 컴포넌트 구현

jeongyou committed:
feat: 카트 등록 완료 확인 버튼 컴포넌트 구현

jeongyou committed:
feat: 카드 등록 완료 페이지에 전달해 줄 카드 정보를 props로 넘겨줌

jeongyou committed:
refactor: 카드 등록 완료 페이지를 location.state 기반으로 수정 및 진입 조건 추가

- 기존 props 전달 방식 제거
- location.state를 통해 카드 정보(cardFirstPartNumbers, cardBrandName) 전달
- 진입 시 location.state가 없으면 /card-register로 리다이렉트되도록 처리
- 잘못된 접근(앞으로 가기, 새로고침 등) 방지

jeongyou committed:
chore: 버튼 컴포넌트 이름 오타 수정 (SubmitButton → ConfirmButton)

jeongyou committed:
dccs: 구현 완료 기능 체크

jeongyou committed:
feat: 카드 정보 입력 유효성 검증 및 확인 버튼 제어 로직 추가

- CardNumbers, CVCNumbers, ExpirationPeriod, CardPassword 컴포넌트에서 error 상태 부모로 끌어올림
- CardRegisterPage에서 모든 입력값의 유효성을 검사하여 SubmitButton 표시 제어
- 입력 중 하나라도 에러가 발생하면 확인 버튼 숨김 처리
- 입력 완료 후 수정 시에도 실시간으로 유효성 반영

jeongyou committed:
refactor: 카드 번호 입력 길이 유효성 검사 로직 분리 및 적용

- CardNumbers 컴포넌트에서 입력값 길이(4자리) 검증 추가
- 에러 배열 체크뿐만 아니라 각 카드 번호 입력이 정확히 4자리 입력되었는지도 검증
- 길이 체크 로직을 isCardNumbersLengthValid 함수로 분리하여 가독성 향상

jeongyou committed:
refactor: CVC 입력 길이 유효성 검사 로직 추가

- CVCNumbers 컴포넌트에서 입력값 길이(3자리) 검증 로직 추가
- 에러 상태뿐만 아니라 입력된 숫자가 정확히 3자리인지 확인하도록 수정
- useEffect에서 error 상태와 cvcNumbers 값을 함께 감시하여 정확한 유효성 검사 적용

jeongyou committed:
docs: 폼 제출 및 상태 관리 기능 구현 완료

jeongyou committed:
chore: 불필요한 console.log 삭제

jeongyou committed:
refactor: CVCNumbers 입력 로직을 useCVCNumbers 커스텀 훅으로 분리

- CVCNumbers 컴포넌트에서 입력값 관리 및 에러 처리 로직을 useCVCNumbers 커스텀 훅으로 분리
- index.tsx는 커스텀 훅 호출 및 View 연결만 담당하도록 역할 분리

jeongyou committed:
refactor: 카드 번호 입력 로직을 useCardNumbers 커스텀 훅으로 분리

- CardNumbers 컴포넌트에서 입력값 관리 및 에러 처리 로직을 useCardNumbers 커스텀 훅으로 분리
- index.tsx는 커스텀 훅 호출 및 View 연결만 담당하도록 역할 분리

jeongyou committed:
chore: 숫자 2 대신 상수를 사용하도록 변경

jeongyou committed:
refactor: 카드 비밀번호 입력 로직을 useCardPassword 커스텀 훅으로 분리

- CardPassword 컴포넌트에서 입력값 관리 및 에러 처리 로직을 useCardPassword 커스텀 훅으로 분리
- index.tsx는 커스텀 훅 호출 및 View 연결만 담당하도록 역할 분리

jeongyou committed:
refactor: 유효기간 입력 로직을 useExpirationPeriod 커스텀 훅으로 분리

- ExpirationPeriod 컴포넌트에서 입력값 관리 및 에러 처리 로직을 useExpirationPeriod 커스텀 훅으로 분리
- 컴포넌트는 입력 포커스 및 블러 UI 조작(handleFocus, handleBlur)만 담당하도록 역할 분리

jeongyou committed:
chore: base url '/' 로 변경

jeongyou committed:
refactor: 이미지 경로 수정

jeongyou committed:
chore: 스토리북에서 이미지 사용을 위한 staticDirs 설정 추가

jeongyou committed:
refactor: 테스트를 위한 data-testid 값 수정

jeongyou committed:
chore: 진짜 public폴더를 찾을 수 있도록 staticDirs 경로 수정

jeongyou committed:
test: 카드 등록 페이지 스토리북 문서화

jeongyou committed:
test: 카드 등록 완료 페이지 스토리북 문서화

jeongyou committed:
test: 카드 번호 입력 컴포넌트 스토리북 문서화, 수정

jeongyou committed:
test: CVC 번호 입력 컴포넌트 스토리북 문서화, 수정

jeongyou committed:
test: 유효기간 입력 컴포넌트 스토리북 문서화, 수정

jeongyou committed:
test: 카드 미리보기 컴포넌트 스토리북 문서화, 수정

jeongyou committed:
test: 카드 비밀 번호 입력 컴포넌트 기본 동작 테스트, 문서화

jeongyou committed:
test: 카드 등록 확인 버튼 컴포넌트 렌더링 테스트 작성, 문서화

jeongyou committed:
test: 카드 정보 입력 확인 버튼 컴포넌트 렌더링 테스트 작성, 문서화

jeongyou committed:
docs: 불필요한 기능 구현 목록 삭제

jeongyou committed:
chore: 불필요한 설정 코드 삭제

jeongyou committed:
chore: vite의 base url을 '/'에서 ''/react-payments/'로 수정

jeongyou committed:
fix: 배포 했을 때 이미지 경로를 못찾는 문제를 해결하기 위해 이미지 경로를 BASE_URL 동적 설정으로 수정

jeongyou committed:
test: 이미지 경로를 못찾는 문제를 해결하기 위해 이미지 경로를 BASE_URL 동적 설정으로 수정

yujo11 commented on src/App.tsx:
```diff
@@ -1,64 +1,20 @@
+          element={<CardRegisterCompletePage />}
+        />
+      </Routes>
+    </BrowserRouter>
```
eact-router-dom 최신 버전에서는 [createBrowserRotuer](https://reactrouter.com/6.30.0/routers/create-browser-router) 사용을 권장하는데요. 

기린이 사용하고 계신 방식 대신 createBrowserRouter를 사용하면 어떤 점이 좋은지 찾아보시고, 적용해보셔도 좋을거 같습니다 ㅎㅎ

jeongyou replied:
오 createBrowserRouter라는 것은 처음 알았습니다...! 그래서 찾아보니까 createBrowserRouter를 사용하면 유지보수성과 확장성이 높아진다고 하더라고요.

현재 방식과 비교하면
현재 방식: 
-  페이지 많아질수록 `<Route>` 주르륵 늘어남 -> 보기 힘들어짐
- 중첩 라우트를 표현하기 어려워 전체 트리 구조를 파악하기 힘들어짐

그래서 현재 방식에서는 결국 페이지가 많아지고 복잡해지면 **JSX 길어짐**, **페이지의 관계 구조 잘 안 보임**, 등등 유지 보수가 힘들어진다는 것을 알았습니다.

반면에 createBrowserRouter은
- 라우트가 객체 배열로 표현되어 트리 구조를 명확하게 볼 수 있다
- routes.ts 같은 파일로 라우터 설정을 분리할 수 있어 관심사 분리가 가능하다

라는 장점이 있다는 것을 알았습니다. 

그래서 최종적으로 createBrowserRouter를 적용하여 라우트 구조를 수정해보았습니다!

관련 커밋: [refactor: BrowserRouter 구조를 createBrowserRouter로 변경하고 routes.tsx로 라우트…](https://github.com/woowacourse/react-payments/pull/470/commits/d34a369a640d2562d05db84779e8d20dc04c888f)

yujo11 replied:
말씀해주신 차이 외에도 Data Loading 기능을 사용할 수 있는지 여부에도 차이가 있으니 참고하시면 좋겠네요 ㅎㅎ https://reactrouter.com/start/data/data-loading

yujo11 commented on src/App.tsx:
```diff
@@ -1,64 +1,20 @@
+        <Route path='/' element={<Navigate to='/card-register' replace />} />
+        <Route path='/card-register' element={<CardRegisterPage />} />
+        <Route
+          path='/card-register-complete'
```
추가로 이렇게 사용되는 `path`들은 상수화를 해주시면 좋을거 같네요

jeongyou replied:
오오 경로를 상수로 관리하면 오타를 방지할 수 있고, 변경 시에도 수정 범위를 줄일 수 있어 유지보수성이 높아지겠네요!! 그래서 상수화를 적용해보았습니다! 앞으로 여러 컴포넌트나 페이지에서 경로를 사용할 때 일관성 있게 관리할 수 있을 것 같습니다~

관련 커밋: [refactor: 경로 문자열(PATH) 상수화](https://github.com/woowacourse/react-payments/pull/470/commits/37bcc171fd4115e6a088a1664845bc5ba82e9e3e)

yujo11 commented on src/App.tsx:
```diff
@@ -1,64 +1,20 @@
-    </Main>
+    <BrowserRouter basename='/react-payments'>
+      <Routes>
+        <Route path='/' element={<Navigate to='/card-register' replace />} />
```
불필요한 redirect로 느껴지는데요. `/`경로에 `<CardRegisterPage>` 컴포넌트를 렌더하지 않고 이렇게 작성하신 이유가 궁금합니다!

jeongyou replied:
저는' /card-register' 페이지를 메인 페이지라고 생각하지 않았습니다...!
그래서 '/'경로는 임시로 '/card-register'로 리다이렉트하도록 설정해두었습니다.

❓나중에 진짜 메인 페이지가 생길 경우를 대비해서 이렇게 구성해보았는데, 혹시 유조는 이 방식에 대해 어떻게 생각하시나요?

yujo11 replied:
저는 불필요한 redirect 라고 생각합니다 `/` 경로를 비워두고 내부적으로도 `/card-register` 경로로 이동시키는게 더 직관적인 동작이 될거 같네요 (개인 취향이 담긴 부분이라 이런 관점이 있다는 것만 알고 넘어가도 괜찮습니다)

yujo11 commented on src/pages/card-register/CardRegisterPage.tsx:
```diff
@@ -0,0 +1,133 @@
+};
+
+const CardRegisterPage = () => {
+  const [cardNumbers, setCardNumbers] = useState<CardNumber>({
+    first: '',
+    second: '',
+    third: '',
+    fourth: '',
+  });
+  const [period, setPeriod] = useState<Period>({
+    month: '',
+    year: '',
+  });
+  const [cvcNumbers, setCvcNumbers] = useState<string>('');
+  const separatorRef = useRef<HTMLDivElement>(null);
+  const [cardFrameColor, setCardFrameColor] = useState('#333333');
+  const [selectedBrand, setSelectedBrand] = useState('');
+  const [isCardNumbersFilled, setIsCardNumbersFilled] = useState(false);
+  const [isPeriodFilled, setIsPeriodFilled] = useState(false);
+  const [passwordNumbers, setPasswordNumbers] = useState('');
+  const [isCVCFilled, setIsCVCFilled] = useState(false);
+  //유효성 검증을 관리하는 상태
+  const [cardNumbersError, setCardNumbersError] = useState(false);
+  const [expirationPeriodError, setExpirationPeriodError] = useState(false);
+  const [cvcError, setCvcError] = useState(false);
+  const [passwordError, setPasswordError] = useState(false);
```
상태가 굉장히 많아졌는데요. 비슷한 역할을 하는 상태끼리 묶어서 관리하는 방법도 있었을거 같아요. 이렇게 모든 상태를 하나하나 개별 상태로 관리하시게 된 배경이 궁금합니다!

jeongyou replied:
사실 처음에는 상태들을 묶어서 관리할 수 있다는 걸 잘 몰랐고, 각 상태를 어떻게 묶어야 할지도 감이 없어서 일단은 각각 따로 useState로 관리하게 되었습니다. 그리고 당시에는 일단 기능 구현이 우선이었기 때문에, 독립적으로 선언해서 빠르게 사용했던 것 같습니다.

사실 지금 생각해보면, 커스텀 훅 내부에서 관리해도 되는 상태들인데 로직을 그렇게 구성하지 못해서 밖에서 중복해서 사용하는 구조가 되었습니다.

이번 리뷰를 통해 비슷한 역할을 하는 상태들을 객체로 묶거나, useReducer를 활용해서 좀 더 구조적으로 관리할 수 있다는 방법을 알게 되었고, 커스텀 훅에서 상태를 내부에서 처리하면 컴포넌트 쪽에서는 관리해야 할 상태 수가 크게 줄어든다는 것도 새롭게 알게 되었습니다!

yujo11 commented on src/pages/card-register/CardRegisterPage.tsx:
```diff
@@ -0,0 +1,133 @@
+          setCardNumbersError={setCardNumbersError}
+        />
+
+        {isAllValid && (
+          <SubmitButton
+            cardFirstPartNumbers={cardNumbers.first}
+            cardBrandName={selectedBrand}
+          >
+            확인
+          </SubmitButton>
+        )}
```
여기도 두 가지 방법이 있었을거 같아요

1. 버튼을 비활성화 해두다가 모든 입력값이 valid한 경우에 활성화
2. 버튼을 노출하지 않다가 모든 입력값이 valid한 경우에 노출

이 중 2번 방법을 선택하셨는데요. 어떤 배경으로 선택하셨는지 궁금합니다

jeongyou replied:
버튼을 비활성화하는 방법은 사용자가 항상 버튼의 존재를 인지할 수 있다는 장점이 있지만, 입력이 완료되지 않아도 버튼이 보이기 때문에 '지금 눌러도 되는 건가?' 하는 혼란을 줄 수 있다고 생각했습니다.
반면, 버튼을 숨기는 방법은 사용자가 입력을 모두 완료하기 전까지 버튼 자체가 보이지 않아, 제출 가능 여부를 보다 직관적으로 전달할 수 있다는 점이 장점이라고 느꼈습니다.

그래서 이번에는 사용자가 입력 상태를 더 명확하게 인식할 수 있도록 2번 방법을 선택했습니다.
입력을 완료하면 버튼이 자연스럽게 나타나고, 다시 수정해서 입력이 부족해지면 버튼이 사라지도록 해서 현재 상태를 바로바로 피드백하려고 했습니다.

❓혹시 유조님은 두 가지 방법 중에 어떤 방식을 더 선호하시나요?

yujo11 replied:
어떤 퍼널을 거치는지, 어떤 상황인지에 따라 달라지겠지만 지금처럼 한 페이지 내에서 모든 input을 받는 구조에서는 1번 방법이 더 직관적으로 느껴져서 개인적으로 선호 하는데요. 정답은 없는 부분이기 때문에 이것 역시 이런 관점이 있다는 것만 알고 넘어가셔도 괜찮을거 같습니다 ㅎㅎ

yujo11 commented on src/pages/card-register/CardRegisterPage.tsx:
```diff
@@ -0,0 +1,133 @@
+            passwordNumbers={passwordNumbers}
+            setPasswordNumbers={setPasswordNumbers}
+            setPasswordError={setPasswordError}
+          ></CardPassword>
```
소소한 부분이지만 self-closing이 가능한 경우 self-closing 태그를 활용해주시면 좋을거 같네요 ㅎㅎ

jeongyou replied:
앗 이부분에 불필요한 닫는 태그가 있었는지 몰랐네요...! 얼른 수정하였습니다~!

관련 커밋: [style: 불필요한 닫는 태그 제거 및 self-closing으로 변경](https://github.com/woowacourse/react-payments/pull/470/commits/83cd5ce6951090579b746e819f30488b73c262e3)

yujo11 commented on src/pages/card-register/CardRegisterPage.tsx:
```diff
@@ -0,0 +1,133 @@
+          <CVCNumbers
+            cvcNumbers={cvcNumbers}
+            setCvcNumbers={setCvcNumbers}
+            onComplete={() => setIsCVCFilled(true)}
```
인라인으로 함수를 넘기는 것과 함수를 선언해서 넘기는 것은 어떤 차이가 있을까요?

jeongyou replied:
헉, 이 개념에 대해서도 이번에 처음 알게 되었습니다!
찾아보니, 인라인으로 함수를 넘기면 렌더링할 때마다 새로운 함수가 생성되어 불필요한 리렌더링이 발생할 수 있고, 반대로 함수를 미리 선언해서 넘기면 매 렌더링마다 참조가 유지되어 성능상 이점이 있다는 것을 알게 되었습니다.

그래서 props로 함수를 넘길 때는 선언해서 사용하는 것이 성능에 더 유리하다는 것도 깨달았습니다. 앞으로는 이런 차이를 생각해서 함수를 선언한 뒤 넘기는 방식을 습관화해보면 좋을것 같습니다.

❓그런데 한 가지 궁금한 점이 있는데요!
저는 인라인으로 작성할 때 코드 가독성이 더 좋다고 느끼는 경우도 있는데, 성능에 큰 차이가 없는 경우라면 인라인으로 넘겨도 괜찮을 것 같은데, 유조는 이 부분에 대해서 어떻게 생각하시나요? 😅

yujo11 replied:
오 잘 찾아주셨군요 기린 멋지네요 ㅎㅎ 👍 👍 👍 

성능과 가독성을 놓고본다면 가독성에 무게를 더 두는 편이라 기린이 말씀해주신 케이스라면 저도 인라인으로 넘길거 같네요 ㅎㅎ

yujo11 commented on src/stories/Components/ConfirmButton.stories.tsx:
이 `.stories` 파일들은 컴포넌트가 선언되는 위치와 가까이 있으면 보기가 더 편할거 같네용 ㅎㅎ (취향에 따라 갈리는 부분이라 변경은 안 해주셔도 됩니다)

jeongyou replied:
오 이 생각은 못했었는데 스토리 파일을 각 컴포넌트 폴더에 두면 보기도 편하고 수정할 때도 편할 것 같네요..! 다음부터 참고하도록 하겠습니다!

yujo11 commented on src/stories/Components/SubmitButton.stories.tsx:
```diff
@@ -0,0 +1,40 @@
+  parameters: {
+    docs: {
+      description: {
+        component: '카드 등록 폼을 모두 작성한 뒤, 제출하는 버튼입니다.',
```
설명들 남겨주신 부분 좋네요 ㅎㅎ 👍 👍 👍

jeongyou replied:
감사합니닷ㅎㅎ

yujo11 commented on src/pages/card-register-complete/CardRegisterCompletePage.tsx:
```diff
@@ -0,0 +1,73 @@
+    if (!cardInfo) {
+      navigate('/card-register', { replace: true });
+    } else {
+      window.history.replaceState(null, '', window.location.href);
```
여기는 `navigate`대신 `replaceState`를 사용하신 이유가 있을까요?

jeongyou replied:
처음에는 사용자가 뒤로 갔다가(뒤로가기) 앞으로 왔을 때(앞으로 가기) 남아 있는 state 문제를 해결하려고 replaceState를 사용했습니다.  그런데 나중에 cardInfo가 없는 경우 바로 /card-register로 리다이렉트하도록 흐름을 변경하면서, 사실상 replaceState는 더 이상 필요 없게 된 것 같아서 이 부분 반영하여 코드 수정하였습니다!

관련 커밋: [refactor: cardInfo 없을 때 리다이렉트 처리로 인해 replaceState 제거](https://github.com/woowacourse/react-payments/pull/470/commits/c223859485595fe36860e3351a3a3d8d48c71842)

yujo11 commented on src/components/CardNumbers/index.tsx:
```diff
@@ -1,68 +1,39 @@
   second: string;
   third: string;
   fourth: string;
 };
```
같은 타입이 여러 곳에 정의되어 있는데요. 한 곳에서 선언해서 import를 통해 같은 출처에서 타입의 형태를 가져올 수 있도록 해보시면 좋을거 같아요

jeongyou replied:
넵 말씀해주신 대로 중복된 타입들을 한 곳에서 관리할 수 있도록 간단하게 types/ 폴더를 만들어 공통 타입을 분리하였고, 각 사용처에서는 해당 타입을 import해서 재사용하도록 반영해보았습니다!
앞으로도 타입을 계속 반복해서 쓰기보다, 같은 도메인의 타입은 하나의 출처에서 관리되도록 구조화하는 것이 좋을 것 같다는 생각이 듭니다~

관련 커밋: [chore: 카드 번호 및 유효기간 타입 정의 파일 분리](https://github.com/woowacourse/react-payments/pull/470/commits/66eea332e3b0c97ff8f126a37e7c75a63255f0df)

❓그런데 이 과정에서 한 가지 궁금증이 생겼습니다.
현재 이 타입들이 컴포넌트, 스토리북, 도메인 로직 등 여러 곳에서 함께 사용되고 있어서 우선은 src/types/ 폴더를 만들어 최상위에 두었는데, 타입을 어떤 기준으로 분리해서 어디에 두는 것이 좋은지에 대한 판단이 잘 서지 않습니다...

예를 들어 컴포넌트에서만 쓰이는 타입이라면 해당 컴포넌트 내부에 두는 게 맞는 건지, 아니면 그래도 공통된 types/ 폴더에 두는 게 일관성 면에서 좋은 건지 고민이 됩니다. 사용하는 위치에 가까운 곳에 두는 게 관리 측면에서는 편할 것 같다는 생각도 들고요.

혹시 타입 파일의 위치나 분리 기준에 대해서 유조만의 규칙이 있는지 궁금합니다!

yujo11 replied:
저는 선언되는 곳과 사용되는 곳이 가까이 있는 것을 선호하는데요.

이런 관점에서 본다면 특정 컴포넌트 하나에서만 쓰이는 타입은 컴포넌트에 위치하거나 컴포넌트 바로 옆에 위치시킬거 같아요

ex)
```
components/Foo/Foo.tsx
components/Foo/Foo.type.ts
```

하지만 앱 전체에 걸쳐 사용되는 타입이라면 src/ 밑에 타입을 선언해두고 사용할거 같네요

ex)
```
src/common/types/common.type.ts
```

yujo11 commented on src/components/CardNumbers/index.tsx:
```diff
@@ -1,68 +1,39 @@
-      return newState;
-    });
-  };
+const CardNumbers = ({
```
로직들을 custom hook으로 분리해주셨는데요.

그렇다면 이 custom hook을 view에서 바로 가져와서 사용하는 방법도 있을거 같아요. 어떤 것을 염두에 두고 지금과 같은 구조를 유지하고 계신지 궁금합니다!

jeongyou replied:
처음에는 유조가 말씀해주신 것처럼 커스텀 훅을 View에서 직접 사용하는 구조도 고민했는데요! 당시에는 커스텀 훅이 도메인 로직과 UI를 분리해주는 역할이라는 개념을 잘 몰랐고, UI와 로직을 명확히 나누기 위해 View와 Container 역할의 컴포넌트를 분리하는 방식으로 쭉 구현했었습니다.
그리고 기존에 작성한 구조를 크게 바꾸지 않고 빠르게 적용할 수 있다는 점도 고려해 지금과 같은 구조를 유지하게 되었습니다 ㅎㅎ

하지만 이후에 다른 미션들을 진행하면서 커스텀 훅이 내부에 도메인 로직을 감추고 있기 때문에, UI 컴포넌트 안에서 바로 호출해 사용해도 관심사가 분리된다는 점을 깨달았습니다. 그래서 이후 미션부터는 해당 방식으로 점점 적용해보고 있습니다!

yujo11 commented on src/components/ConfirmButton/ConfirmButton.tsx:
```diff
@@ -0,0 +1,44 @@
+  margin: 0 auto;
+  display: block;
+  border-radius: 5px;
+  font-family: Noto Sans KR;
```
이런 font-family의 경우 전역 스타일로 적용을 해줘도 좋을거 같네요

jeongyou replied:
말씀해주신 대로 font-family를 전역으로 적용하는 것이 스타일 일관성과 관리 측면에서 더 좋다고 생각하여, GlobalStyle을 통해 기본 font-family로 Noto Sans KR을 설정해주었습니다!

관련 커밋: [style: font-family 전역 적용 및 컴포넌트별 중복 제거](https://github.com/woowacourse/react-payments/pull/470/commits/8ccde62df11d087cdddde3bde4972687baaaa473)

yujo11 commented on src/pages/card-register-complete/CardRegisterCompletePage.tsx:
```diff
@@ -0,0 +1,73 @@
+  width: 376px;
+  height: auto;
+  min-height: 100dvh;
+  background-color: #f9f9f9;
```
자주 사용되는 색상들도 변수로 정의해서 관리하면 전체 애플리케이션에서 일관된 스타일을 적용하는데 도움이 될거 같아요

jeongyou replied:
헉 그렇네요...! 자주 사용하는 색상을 상수로 관리하면 스타일 일관성을 유지하기 좋을 것 같습니다. 그래서 colors.ts 파일을 추가하고, background-color 및 카드사 색상 등을 상수로 정의하여 적용해보았습니다!

관련 커밋: [style: colors.ts 추가 및 색상 상수화](https://github.com/woowacourse/react-payments/pull/470/commits/cb80aa0647ae4dc58175fa1864f24433d5ae21b0)

yujo11 commented on src/components/common/SingleButton.tsx:
```diff
@@ -0,0 +1,35 @@
+import styled from '@emotion/styled';
+
+interface SingleButtonProps {
+  type?: 'submit' | 'reset' | 'button';
+  children: React.ReactNode;
+  onClick?: () => void;
+  disabled?: boolean;
+  className?: string;
+}
```
애플리케이션 전체에 걸쳐 html button을 대신해 사용하게 될 컴포넌트로 보이는데요. .그렇다면 html 버튼의 기능이 가질 수 있는 attribute를 모두 받을 수 있게 해주시면 보다 유연한 사용이 가능할거 같아요

jeongyou replied:
말씀해주신 대로 SingleButton이 다양한 상황에서 사용될 수 있기 때문에, HTML <button> 태그의 기본 속성들을 모두 지원할 수 있도록 React.ButtonHTMLAttributes<HTMLButtonElement>를 상속받는 형태로 수정하였습니다! 

관련 커밋: [refactor: SingleButton 컴포넌트에 Button HTML 기본 속성 지원 추가](https://github.com/woowacourse/react-payments/pull/470/commits/2adef125c0442a405b2c52c30ba205fdf3edb876)

yujo11 reviewed (changes requested):
안녕하세요 기린 ~~ 해외 잘 다녀왔습니다 ㅎㅎ 빠르게 리뷰 요청 주셨는데 리뷰 남기는게 조금 늦어졌네요 다음 번 리뷰는 최대한 빠르게 진행할 수 있도록 하겠습니다!

이번 리뷰는 전체적으로 코드를 보면서 작성 배경에 대한 질문을 많이 남겼는데요. 확인해주시면 좋을거 같네요!

---

>1. 기본 컴포넌트 생성의 필요성

기린이 말씀해주신 SingleButton 컴포넌트는 기린의 애플리케이션 내에서 html의 button element를 대신해서 사용 될 컴포넌트로 보이는데요. 전체 코드에서 공통 스타일과 동작을 포함하여 일관된 모영과 동작을 보장할 수 있을거 같아요.

지금 SingleButton 컴포넌트를 다시 래핑해서 사용한 곳들을 보면 backgroundColor, color, border-radius 등을 사용하는 곳에서 결정하고 있는데요. SingleButton 컴포넌트에 이러한 부분들을 정의하고 사용한다면 손쉽게 일관성 있는 코드를 작성할 수 있을거 같아요 

ex)
```ts
interface Button {
  size: "small" | "default" | "large"
  rounded: boolean
  color: "primary" | "default" | "danger"
}
```

잘 만들어진 디자인 시스템들의 이런 html element들에 대응하는 컴포넌트들은 어떤 형태로 만들어져 있고, 어떤 prop을 이용할 수 있는지 확인해보시는 것도 도움이 되실거 같습니다!

- https://ant.design/components/button

>2. 커스텀 훅 사용

커스텀 훅을 생각해 볼 때 두 가지 관점에서 접근해 볼 수 있을거 같은데요. 하나는 정말 특정한 용도로만 사용되는 hook이고, 하나는 마치 util 함수처럼 애플리케이션 전체에 걸쳐 사용할 수 있는 hook이 있을거에요.

매번 완전히 새로운 훅을 만드는 대신에 재사용 가능한 기본 훅을 만들고, 기본 훅을 바탕으로 다른 훅들을 만들어나간다면 이런 부담이 조금은 줄어들 수 있을거 같네요

현재 코드에서는 각 입력 필드마다 서로 다른 유효성 검사 규칙을 가지고 있어 별도로 만드는 것도 일리가 있어보이구요 ㅎㅎ

[질문]  UI와 Domain 코드를 분리하는 장점을 체감하셨다고 하셨는데 어떤 부분이 장점으로 느껴졌고, UI와 Domain 로직이 결합되어 있으면 어떤 점이 안 좋다고 생각하시는지 궁금합니다

>3. 입력 필드 자동 이동

장단점이 있는 부분인거 같아요. 말씀해주신대로 구현한다면 오류 수정 시 번거로움이 줄어들겠지만 자동으로 이동한다면 정상적인 유저 플로우에서는 더 편하게 입력이 가능할거 같은데요. 트레이드오프가 각각 존재하고 있어 정답은 없다고 생각해요. 조금 더 나은 대안을 생각해본다면 손은 많이 가겠지만 오류가 발생했을 때는 필드를 유지하고, 오류가 발생하지 않았을 때는 다음 step의 컴포넌트로 포커스가 이동하는 방법도 생각해 볼 수 있겠네요 ㅎㅎ

저의 경우에는 애플리케이션의 요구사항에 따라 다르게 구현하는데 mobile에서는 자동으로 이동하는 방식을, 웹에서는 자동으로 이동하지 않는 방식을 많이 사용했던거 같네요 ㅎㅎ

>4. react-router를 이용한 페이지 관리

유저의 이동을 직접적으로 막는 방법이 좋은 방법이다라고 할 수는 없겠지만 상황에 따라 필요한 방법인데요. 현재 작성해주신 내용도 충분히 일리가 있다고 생각해요. 왜 redirect 되는지 유저에게 알려주는 방법도 좋은 아이디어라는 생각이 드네요! react-rotuer-dom에 `useBlocker`라는 hook이 있는데 이 hook에 대해서도 한번 찾아보시면 좋을거 같습니다!

저의 경우에는 페이지가 이동 되기 전에 alert message와 함께 차단하는 것을 선호합니다 ㅎㅎ

---

미션 진행하느라 고생 많으셨습니다 기린 ~~ 👍 👍 👍

jeongyou committed:
style: 불필요한 닫는 태그 제거 및 self-closing으로 변경

jeongyou committed:
refactor: BrowserRouter 구조를 createBrowserRouter로 변경하고 routes.tsx로 라우트 설정 분리

jeongyou committed:
refactor: cardInfo 없을 때 리다이렉트 처리로 인해 replaceState 제거
- 기존에는 뒤로 가기/앞으로 가기 시 남아있는 데이터 문제를 해결하기 위해 replaceState를 사용했으나
- cardInfo 없으면 바로 리다이렉트하는 로직으로 변경되면서 필요 없어져 삭제

jeongyou committed:
refactor: SingleButton 컴포넌트에 Button HTML 기본 속성 지원 추가

jeongyou committed:
refactor: 경로 문자열(PATH) 상수화

jeongyou committed:
style: font-family 전역 적용 및 컴포넌트별 중복 제거
- GlobalStyle에서 Noto Sans KR 지정
- 컴포넌트 내 font-family 속성 삭제

jeongyou committed:
style: colors.ts 추가 및 색상 상수화

- 공통 배경색, 텍스트 색상, 카드사 색상 등 상수로 관리

jeongyou committed:
chore: 카드 번호 및 유효기간 타입 정의 파일 분리

jeongyou committed:
refactor: 분리된 타입을 기존 컴포넌트에 적용

jeongyou committed:
test: 분리된 타입을 스토리북에 적용

jeongyou commented:
안녕하세요 유조! 꼼꼼한 리뷰 감사합니다!
리뷰를 통해 기존 코드에서 미처 고려하지 못했던 개념을 다시 점검할 수 있었고, 더 나은 방향으로 개선할 수 있는 방법도 고민해 볼 수 있어서 정말 많은 도움이 되었습니다.
좋은 피드백 주셔서 다시 한번 감사드립니다! 🙌

**각 리뷰 코멘트에 대해 수정 사항을 반영했고, 해당 댓글에 코멘트를 남겨두었습니다!**

그리고 처음에 질문 주셨던 
>UI와 Domain 코드를 분리하는 장점을 체감하셨다고 하셨는데 어떤 부분이 장점으로 느껴졌고, UI와 Domain 로직이 결합되어 있으면 어떤 점이 안 좋다고 생각하시는지 궁금합니다.

라는 질문도 함께 고민해보았습니다.

처음에는 UI 코드 안에 도메인 로직(입력값 처리, 유효성 검사 등)을 같이 넣어서 작성했었는데, 컴포넌트가 점점 길어지고 복잡해지면서 어떤 코드가 UI를 위한 건지, 어떤 게 로직인지를 구분하기가 어려워졌습니다. 특히 에러 상태나 입력 완료 여부처럼 로직이 점점 늘어날수록 UI 관련 JSX와 로직 코드가 섞이면서 가독성이 많이 떨어졌습니다.

그런데 커스텀 훅을 만들어서 도메인 로직을 분리해보니까, 컴포넌트는 props만 받아서 렌더링에만 집중할 수 있게 되고,
로직은 로직대로 독립적으로 관리되면서 코드가 훨씬 읽기 쉬워졌습니다!

아! 그리고 마지막으로, 아래 리뷰에 대해서도 따로 말씀드리고 싶습니다.
> 상태가 굉장히 많아졌는데요. 비슷한 역할을 하는 상태끼리 묶어서 관리하는 방법도 있었을거 같아요. 이렇게 모든 상태를 하나하나 개별 상태로 관리하시게 된 배경이 궁금합니다!

다만 이 구조를 현재 프로젝트에 적용하려면 커스텀 훅과 컴포넌트 로직 전반을 거의 새롭게 수정해야 해서, 우선은 먼저 수정이 완료된 부분에 대해 리뷰를 받고, 나중에 시간이 날 때 차근차근 적용해보려고 합니다.

감사합니다!

yujo11 commented on src/routes/routes.tsx:
```diff
@@ -0,0 +1,23 @@
+import CardRegisterPage from '../pages/card-register/CardRegisterPage';
+import CardRegisterCompletePage from '../pages/card-register-complete/CardRegisterCompletePage';
+
+export const router = createBrowserRouter(
```
👍 👍 👍

yujo11 commented on src/pages/card-register/CardRegisterPage.tsx:
```diff
@@ -71,7 +71,7 @@ const CardRegisterPage = () => {
             setPasswordNumbers={setPasswordNumbers}
             setPasswordError={setPasswordError}
-          ></CardPassword>
+          />
```
👍 👍 👍

yujo11 commented on src/components/common/SingleButton.tsx:
```diff
@@ -1,11 +1,8 @@
-interface SingleButtonProps {
-  type?: 'submit' | 'reset' | 'button';
+interface SingleButtonProps
+  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
```
👍 👍 👍

yujo11 commented on src/routes/routes.tsx:
```diff
@@ -0,0 +1,29 @@
+import CardRegisterPage from '../pages/card-register/CardRegisterPage';
+import CardRegisterCompletePage from '../pages/card-register-complete/CardRegisterCompletePage';
+
+export const PATH = {
+  ROOT: '/',
+  CARD_REGISTER: '/card-register',
+  CARD_REGISTER_COMPLETE: '/card-register-complete',
+};
```
저도 같은 리뷰를 남기려고 했는데 코드래빗이 잡아줬군요 ㅎㅎ

<img width="1645" alt="image" src="https://github.com/user-attachments/assets/0e926d1a-2ec6-43c5-bbdd-75eafbf15fae" />

yujo11 commented on src/components/CardNumbers/CardNumbersView.tsx:
```diff
@@ -2,13 +2,7 @@ import React from 'react';
 import FormLabel from '../common/FormLabel';
 import InputTexts from '../common/InputTexts';
-
-type CardNumber = {
-  first: string;
-  second: string;
-  third: string;
-  fourth: string;
-};
```
타입의 출처가 여러군데가 되면 타입시스템의 이점을 잘 누리지 못 하게 되는데요. 동알한 타입에 대해서는 단일 출처를 유지한다면 타입이 변경되었을 때 쉽게 알아채고 대응할 수 있을거에요. 이런 부분도 앞으로 타입스크립트를 사용하면서 신경 써보시면 좋을거 같아요


ex) Case 1. 타입을 각각 따로 사용하는 경우 CardBrandName이 변경되어도 타입시스템에서 에러를 감지하지 못 함
```ts
type CardBrandName = "BC" | "SAMSUNG"

const CardBrandInfoMape : Record<"BC" | "SAMSUNG", { color: string }> = {
   BC: { color: "red" },
   SAMSUNG: { color: "blue" }
}
```

ex) Case 2. 단일 출처의 타입을 사용하는 경우 CardBrandName이 변경되었을 때 `CardBrandInfoMap`도 변경되어야 한다는걸 쉽게 알 수 있음
```ts
type CardBrandName = "BC" | "SAMSUNG"

const CardBrandInfoMape : Record<CardBrandName, { color: string }> = {
   BC: { color: "red" },
   SAMSUNG: { color: "blue" }
}
```

yujo11 reviewed (approved):
안녕하세요 기린 ~~ 

다음 미션도 병행하며 정신 없으셨을텐데 코멘트들 꼼꼼하게 확인하고 반영해주셔서 감사합니다 ㅎㅎ

이전 코멘트들에 기린이 남겼던 질문들에 대한 답변과 더불어 몇 가지 소소한 코멘트들을 추가로 남겼으니 확인해보시면 좋을거 같네요!

이번 미션은 여기서 이만 마쳐도 좋을거 같습니다 ㅎㅎ 미션 진행하느라 고생 많으셨습니다 기린! 다음 미션도 응원하겠습니다~~!! 👍 👍 👍
